src/rust/src/contract.rs

```diff
@@ -210,30 +210,40 @@ fn encode_diagnostic_events(events: &Vec<DiagnosticEvent>) -> Vec<RustBuf> {
 
 fn extract_ledger_effects(
     entry_changes: Vec<LedgerEntryChange>,
-) -> (Vec<RustBuf>, Vec<ReadOnlyBump>) {
+) -> Vec<RustBuf> {
     let mut modified_entries = vec![];
-    let mut read_only_bumps = vec![];
 
     for change in entry_changes {
-        if change.read_only {
-            // Only return expiration bumps for read-only entries (if any).
-            if let Some(expiration_change) = change.expiration_change {
-                if expiration_change.new_expiration_ledger > expiration_change.old_expiration_ledger
-                {
-                    read_only_bumps.push(ReadOnlyBump {
-                        ledger_key: change.encoded_key.into(),
-                        min_expiration: expiration_change.new_expiration_ledger,
-                    });
-                }
-            }
-        } else {
+        // Extract ContractCode and ContractData entry changes first
+        if !change.read_only {
             if let Some(encoded_new_value) = change.encoded_new_value {
                 modified_entries.push(encoded_new_value.into());
             }
         }
+
+        // Check for ExpirationEntry changes
+        if let Some(expiration_change) = change.expiration_change {
+            if expiration_change.new_expiration_ledger > expiration_change.old_expiration_ledger
+            {
+                // entry_changes only encode LedgerEntry changes for ContractCode and ContractData
+                // entries. Changes to ExpirationEntry are recorded in expiration_change, but does
+                // not contain an encoded ExpirationEntry. We must build that here.
+                let hash_bytes: [u8; 32] = expiration_change.key_hash.try_into().unwrap();
```

Comment:
I don't think this `unwrap` and the one on line 240 are safe. @graydon if these fail would it lead to an internal error?

The error should be propogated up like we do when `non_metered_xdr_from_cxx_buf` is called.

Reply:
Yeah this will panic if it's wrong.